### DISCUSSION TITLE
Can’t see logs of levels `silly`, `verbose`, `debug`, `fine` in Nest

### DISCUSSION THREAD

**Discussion by tukusejssirs** [jmcdo29/ogma#1426] [Q&A]
I use Ogma in Nest. When I try to log messages of `silly`, `verbose`, `debug`, `fine` levels, nothing is output.

In the module, I have this:

```ts
@Module({
  imports: [
    OgmaModule.forRoot({
      interceptor: {
        http: FastifyParser
      },
      service: {
        application: 'app-name',
        color: true,
        json: false,
        // It does not matter what log level I set in here.
        logLevel: 'ALL'
      }
    }),
    OgmaModule.forFeature(SomeService.name)
  ]
})
export class SomeModule {}
```

In the service, I have this:

```ts
@Injectable()
export class SomeService {
  constructor(@OgmaLogger(SomeService.name) private readonly logger: OgmaService) {}

  someMethod() {
    // These logs won't show up.
    this.logger.silly('some log')
    this.logger.verbose('some log')
    this.logger.debug('some log')
    this.logger.fine('some log')

    // These logs work as expected.
    this.logger.info('some log')
    this.logger.log('some log')
    this.logger.warn('some log')
    this.logger.error('some log')
    this.logger.fatal('some log')
  }
}
```

Is there something I am missing or is it a bug?

**Comment by jmcdo29** [comment]
That's super interesting. Got a reproduction you can share with it? Everything looks fine

**Reply by tukusejssirs** [reply]
No reproduction as of yet.

Another thing I’ve just noticed: `FastifyParser`. It is a TCP microservice; should I set `interceptor.http` to something else?

I created the microservice using:

```ts
import {NestFactory} from '@nestjs/core'
import {MicroserviceOptions, Transport} from '@nestjs/microservices'

const app = await NestFactory.createMicroservice<MicroserviceOptions>(SomeModule, {
  options: {
    host: '127.0.0.1',
    port: 1234
  },
  transport: Transport.TCP
})
```

**Reply by jmcdo29** [reply]
If this is a TCP server, you should use [`rpc: TcpParser`](https://ogma.jaymcdoniel.dev/docs/nestjs/rpc/tcp)

**Reply by tukusejssirs** [reply]
Okay, no change between setting it to `rpc: TcpParser` or `http: FastifyParser`: both work the same way (some log levels work, others don’t).

I’ll try and create a reproduction.

**Reply by tukusejssirs** [reply]
I tried to create a minimal Nest microservice, but all log levels work in there as expected.

There must be something else going on …

**Reply by jmcdo29** [reply]
If you can provide access to the project where the error happens I'd be happy to try and take a look as well

**Reply by tukusejssirs** [reply]
I would provide you access if only it wouldn’t be a work project.

I’ll try to reproduce it without the project. Any idea is welcome! :wink:

**Reply by tukusejssirs** [reply]
After adding some `console.log()`s to the Ogma JS code, I have found out that log messages are correctly process through all the methods (like `OgmaService.printMessage()`, `Omga.debug()`, `Ogma.printMessage()`). To me, it looks like the log level the default value (`INFO`), regardless of what I define in `OgmaModule.forRoot()`.

So, now the question is why does it happen? Could it be related (somehow) to TS config, Nest config, Nx config (which uses Webpack)?

**Reply by jmcdo29** [reply]
Hmm, well, the `OgmaProvider` which ends up creating the initial `Ogma` instance [should respect what options are passed](https://github.com/jmcdo29/ogma/blob/main/packages/nestjs-module/src/ogma.provider.ts#L20) and the class itself should only [re-write the logLevel back to the default if the provided value doesn't match an expected value](https://github.com/jmcdo29/ogma/blob/main/packages/logger/src/logger/ogma.ts#L38) but if that were the case there should also be a warning in the logs that said the passed value wasn't a valid one.

**Reply by tukusejssirs** [reply]
I have added `console.log(options) into `OgmaModule.rorRoot()` method and it outputs correct log level (the one I provided).

[`createOgmaProvider()`](https://github.com/jmcdo29/ogma/blob/main/packages/nestjs-module/src/ogma.provider.ts#L20) and `Ogma` both use the log level I provided. However, messages above `INFO` level are not logged.

Just a note: currently for simplicity I test this using `ALL` level, however, it seems to work (apart from the actual logging of messages above `INFO` level).

I’ll try to find some time for the debugging tomorrow (CEST timezone).

Thanks for your time and help! :pray:

**Reply by tukusejssirs** [reply]
I have some progress, albeit I haven’t found the reason.

---

Notes:

- I’ve debugged this by modifying the compiled sources installed via NPM.
- Ogma package versions:
   - `@ogma/logger@2.4.1`;
   - `@ogma/nestjs-module@4.1.1`;
   - `@ogma/platform-fastify@4.0.0`;
   - `@ogma/platform-tcp@4.0.0`.
- While debugging, I noticed that `context` and `application` values from message command (e.g. `this.logger.info(msg, {application: 'ogma-test-app', context: 'ogma-test-context'}`) are ignored, i.e. the app/context value from Nest module is always used.
- I could go backwards and find out why the log level is always `INFO` in `printMessage()` in `@ogma/logger`. I think you could more easily find the next clue to the puzzle. :wink:
- This can be simplified using ternary operator. Any reason against using it?

https://github.com/jmcdo29/ogma/blob/882c3023077790df245834c3ba4b293af30f5179/packages/logger/src/logger/ogma.ts#L68-L73

```js
const logString = this.options.json ? this.formatJSON(message, options) : this.formatStream(message, options);
```

---

```ts
// Global config
// Note: This is correct (as set in my Nest module).
// Note: In does not matter what `logLevel` I choose, all work here as expected, but in `printMessage()` (see below), I always get `INFO` log level.
// Note: This is output when add the following (commented out) code to `createOgmaProvider()` in `node_modules/@ogma/nestjs-module/src/ogma.provider.js` (all I did is assigned the returned value to a variable, output some data and return the variable value).
// function createOgmaProvider(options) {
//   const ogma = new logger_1.Ogma(Object.assign(Object.assign({}, options), { application: (options === null || options === void 0 ? void 0 : options.application) || 'Nest' }));
//   // Note: `light-indicator` is the app name as configured in the Nest module.
//   if (options.application === 'light-indicator') {
//     console.log('aaaa : provider/createOgmaProvider :', {
//       'options': options,
//       'ogma.options.logLevel': ogma.options.logLevel
//     })
//   }
//   return ogma;
// }
aaaa : provider/createOgmaProvider : {
  options: {
    application: 'light-indicator',
    color: true,
    json: false,
    logLevel: 'ALL'
  },
  'ogma.options.logLevel': 'ALL'
}

// Messages are correctly processed, however `this.options.logLevel` is incorrect (it should be set to `ALL`, as per the global config above)
// Note: This is output when we add the following (commented out) code at the beginning of `printMessage()` in `node_modules/@ogma/logger/src/logger/ogma.js`
// // Note: `light-indicator` is the app name as configured in the Nest module.
// if (options.context === 'SomeService') {
//   console.log('aaaa : logger/printMessage :', {
//     'options': options,
//     message,
//     'this.options.logLevel': this.options.logLevel
//   })
// }
aaaa : logger/printMessage : {
  options: {
    level: 0,
    formattedLevel: '\x1B[35m[SILLY]\x1B[0m',
    context: 'SomeService'
  },
  message: 'test msg : 2022-10-27T14:06:18.988Z 1',
  'this.options.logLevel': 'INFO'
}
aaaa : logger/printMessage : {
  options: {
    level: 1,
    formattedLevel: '\x1B[32m[FINE] \x1B[0m',
    context: 'SomeService'
  },
  message: 'test msg : 2022-10-27T14:06:18.988Z 2',
  'this.options.logLevel': 'INFO'
}
aaaa : logger/printMessage : {
  options: {
    level: 2,
    formattedLevel: '\x1B[34m[DEBUG]\x1B[0m',
    context: 'SomeService'
  },
  message: 'test msg : 2022-10-27T14:06:18.988Z 3',
  'this.options.logLevel': 'INFO'
}
aaaa : logger/printMessage : {
  options: {
    level: 1,
    formattedLevel: '\x1B[32m[FINE] \x1B[0m',
    context: 'SomeService'
  },
  message: 'test msg : 2022-10-27T14:06:18.988Z 4',
  'this.options.logLevel': 'INFO'
}
aaaa : logger/printMessage : {
  options: {
    level: 3,
    formattedLevel: '\x1B[36m[INFO] \x1B[0m',
    context: 'SomeService'
  },
  message: 'test msg : 2022-10-27T14:06:18.988Z 5',
  'this.options.logLevel': 'INFO'
}

// The following logs are output (as you can see) correctly, as they are all above `INFO` level
[2022-10-27T14:06:18.991Z] [INFO]  [hostname] [app-name] [25]  [SomeService] test msg : 2022-10-27T14:06:18.988Z 5
aaaa : logger/printMessage : {
  options: {
    level: 3,
    formattedLevel: '\x1B[36m[INFO] \x1B[0m',
    context: 'SomeService'
  },
  message: 'test msg : 2022-10-27T14:06:18.988Z 6',
  'this.options.logLevel': 'INFO'
}
[2022-10-27T14:06:18.991Z] [INFO]  [hostname] [app-name] [25]  [SomeService] test msg : 2022-10-27T14:06:18.988Z 6
aaaa : logger/printMessage : {
  options: {
    level: 4,
    formattedLevel: '\x1B[33m[WARN] \x1B[0m',
    context: 'SomeService'
  },
  message: 'test msg : 2022-10-27T14:06:18.988Z 7',
  'this.options.logLevel': 'INFO'
}
[2022-10-27T14:06:18.991Z] [WARN]  [hostname] [app-name] [25]  [SomeService] test msg : 2022-10-27T14:06:18.988Z 7
aaaa : logger/printMessage : {
  options: {
    level: 5,
    formattedLevel: '\x1B[31m[ERROR]\x1B[0m',
    context: 'SomeService'
  },
  message: 'test msg : 2022-10-27T14:06:18.988Z 8',
  'this.options.logLevel': 'INFO'
}
[2022-10-27T14:06:18.992Z] [ERROR] [hostname] [app-name] [25]  [SomeService] test msg : 2022-10-27T14:06:18.988Z 8
aaaa : logger/printMessage : {
  options: {
    level: 6,
    formattedLevel: '\x1B[41m\x1B[37m\x1B[4m[FATAL]\x1B[0m',
    context: 'SomeService'
  },
  message: 'test msg : 2022-10-27T14:06:18.988Z 9',
  'this.options.logLevel': 'INFO'
}
[2022-10-27T14:06:18.992Z] [FATAL] [hostname] [app-name] [25]  [SomeService] test msg : 2022-10-27T14:06:18.988Z 9

// This is a method I have in `SomeService`
ogmaTest(msg: string): void {
   // These logs won't show up.
   this.logger.silly(`${msg} 1`, {application: 'ogma-test', context: 'ogma-test'})
   this.logger.verbose(`${msg} 2`, {application: 'ogma-test', context: 'ogma-test'})
   this.logger.debug(`${msg} 3`, {application: 'ogma-test', context: 'ogma-test'})
   this.logger.fine(`${msg} 4`, {application: 'ogma-test', context: 'ogma-test'})

   // These logs work as expected.
   this.logger.info(`${msg} 5`, {application: 'ogma-test', context: 'ogma-test'})
   this.logger.log(`${msg} 6`, {application: 'ogma-test', context: 'ogma-test'})
   this.logger.warn(`${msg} 7`, {application: 'ogma-test', context: 'ogma-test'})
   this.logger.error(`${msg} 8`, {application: 'ogma-test', context: 'ogma-test'})
   this.logger.fatal(`${msg} 9`, {application: 'ogma-test', context: 'ogma-test'})
}

// This is how I call that method in another method
this.ogmaTest(`test msg : ${new Date().toISOString()}`)
```

**Reply by jmcdo29** [reply]
This is all on a class with `@OgmaLogger()` vs just `private readonly logger: Logger` with `Logger` being from `@nestjs/common`, right?

**Reply by tukusejssirs** [reply]
Yes, exactly as I described in the OP, and, of course, I have the following in the service:

```ts
@Injectable()
export class SomeService {
  constructor(
    @OgmaLogger(SomeService.name) private readonly logger: OgmaService
  ) {}
}
```

> with Logger being from `@nestjs/common`, right?

I am puzzled. Why would I need to import `@nestjs/common` when I use `@ogma/logger` + `@gma/nestjs-module`? No, `@nestjs/common` is not loaded anywhere in this microservice (only `import {OgmaLogger, OgmaService} from '@ogma/nestjs-module'`).

**Reply by jmcdo29** [reply]
You don't need to import `@nestjs/common`. That's what I wanted to make sure _wasn't_ happening

**Reply by tukusejssirs** [reply]
Oops, I’ve misread and misinterpreted your previous comment, sorry. :man_facepalming:

**Reply by jmcdo29** [reply]
You don't happen to somehow call `OgmaModule.forRoot()` twice in your server do you? I'm wracking my brain here trying to think of what could be happening

**Reply by tukusejssirs** [reply]
Well, no, not in `SomeModule` module. However, I have a Nest library that I import in `SomeModule` and that library also imports `OgmaModule.forRoot()` and `OgmaModule.forFeature()`. Both are in an Nx monorepo.

Could that be related? :thinking:

**Reply by jmcdo29** [reply]
I think so. Under the hood, `OgmaModule.forRoot()` ends up calling to `OgmaCoreModule.forRoot()` which sets up a couple of global providers. If `forRoot()` gets called to multiple times, it could be overwriting the providers from earlier and no longer persisting the options from the first call

**Reply by tukusejssirs** [reply]
Okay, so how should I deal with this? Should I simply remove `OgmaModule.forRoot()` from the library and keep there only `OgmaModule.forFeature()`? I’ve just tried that out, but it still does not work.

Just I note: I use an env var (`LOG_LEVEL`) which is used to define the log level throughout all apps (gateway and microservices), thus it should be the same. I’ve set it to `ALL`, but I don’t get any messages of `silly`, `verbose`, `debug`, `fine` level.

**Reply by jmcdo29** [reply]
Hmm, that's super strange. Really wish there was a way to replicate this, cause it _sounds_ like a possible configuration issue, but it's hard to say _why_, y'know?

**Comment by tukusejssirs** [comment]
> If this is a TCP server, you should use [`rpc: TcpParser`](https://ogma.jaymcdoniel.dev/docs/nestjs/rpc/tcp)

Just to make sure: in TCP microservices I should use `rpc: Parser` and in gateway (regular Nest app with Fastify) I should use `http: FastifyParser`, right?

> Hmm, that's super strange. Really wish there was a way to replicate this, cause it _sounds_ like a possible configuration issue, but it's hard to say _why_, y'know?

It does not matter if I log the messages in a microservice or in the gateway, it always defaults to `INFO` log level and thus I get only a subset of logs.

Yeah, it seems to be a configuration issue, as I am still quite new to TS/Nest/Nx and evething in between (I work with these for about a year and I have only self-taught myself from online resources).

I understand it is hard to debug without a reproduction. I’ll try to find some time to reproduce it again, however, the thing is I have no idea what to reproduce and is not needed. The code is quite complex, thus there might be to many stuff that (potentially) could interfere.

---

Anyway, I removed `OgmaModule.forRoot()` from the library, however, it didn’t solve the issue.

**Reply by jmcdo29** [reply]
Quick question: are you using a hardcoded string for the log level or is it a env variable or ternary expression here?

---

> Just to make sure: in TCP microservices I should use `rpc: Parser`

Yes

> in gateway (regular Nest app with Fastify) I should use http: FastifyParser, right?

Yes

> It does not matter if I log the messages in a microservice or in the gateway, it always defaults to INFO log level and thus I get only a subset of logs.

And these are separate applications, so they shouldn't be sharing configuration which makes this even weirder

**Reply by tukusejssirs** [reply]
> Quick question: are you using a hardcoded string for the log level or is it a env variable or ternary expression here?

For debugging purposes, I hard-coded it to `ALL` string, otherwise I use an env variable defined in a file, loaded at gateway startup, validated if it contains a valid value, defaults to `WARN` if not provided or if invalid value provided. No ternary expression is used in this case.

Thanks for the answers. :pray:
